File: Core/Models/CoachExtras.swift
import Foundation

// Embedded enums + structs that hang off Coach for the federation-grade
// profile redesign (Stage 1.6). Mirrors the Athlete dossier pattern — no new
// Repository methods, keeps the Android port mechanical.

public enum CoachLevel: String, Codable, CaseIterable, Sendable, Hashable {
    case assistant, junior, senior, head, national, international

    public var labelKey: String { "coach.level.\(rawValue)" }
}

public enum CoachLicenseLevel: String, Codable, CaseIterable, Sendable, Hashable {
    case poom2, poom1, dan1, dan2, dan3, dan4, dan5plus

    public var labelKey: String { "coach.license.\(rawValue)" }
}

public enum CoachSpecialisation: String, Codable, CaseIterable, Sendable, Hashable {
    case kyorugi, poomsae, technical, fitness, sparring, multi

    public var labelKey: String { "coach.specialisation.\(rawValue)" }

    public var systemIcon: String {
        switch self {
        case .kyorugi: "figure.martial.arts"
        case .poomsae: "figure.mind.and.body"
        case .technical: "figure.taichi"
        case .fitness: "figure.run"
        case .sparring: "figure.boxing"
        case .multi: "circle.grid.3x3.fill"
        }
    }
}

public enum CoachEmploymentStatus: String, Codable, CaseIterable, Sendable, Hashable {
    case active, leave, transferred, retired, suspended

    public var labelKey: String { "coach.employment.\(rawValue)" }
}

public enum CoachProgramStatus: String, Codable, CaseIterable, Sendable, Hashable {
    case none, candidate, supportStaff, member, leadStaff

    public var labelKey: String { "coach.program.\(rawValue)" }
}

/// Cross-federation ranking snapshot for a coach. Nil fields render as "—".
public struct CoachRanking: Codable, Hashable, Sendable {
    /// Internal rank inside SSDC (across all coaches).
    public var club: Int?
    /// UAE Taekwondo Federation coach standing.
    public var uae: Int?
    /// World Taekwondo recognition tier label (e.g. "Class A", "Class B").
    /// String — WT publishes tiers, not numeric ranks.
    public var worldTier: String?
    public var asOf: Date

    public init(club: Int? = nil, uae: Int? = nil, worldTier: String? = nil, asOf: Date = Date()) {
        self.club = club
        self.uae = uae
        self.worldTier = worldTier
        self.asOf = asOf
    }
}

File: flutter_app/lib/core/models/coach_extras.dart
// Port of Core/Models/CoachExtras.swift.
// Embedded enums + structs that hang off Coach for the federation-grade
// profile redesign (Stage 1.6). Mirrors the Athlete dossier pattern — no new
// Repository methods, keeps the Android port mechanical.

enum CoachLevel {
  assistant,
  junior,
  senior,
  head,
  national,
  international;

  String get labelKey => 'coach.level.$name';

  static CoachLevel fromJson(String raw) => CoachLevel.values.firstWhere(
    (c) => c.name == raw,
    orElse: () => CoachLevel.assistant,
  );
}

enum CoachLicenseLevel {
  poom2,
  poom1,
  dan1,
  dan2,
  dan3,
  dan4,
  dan5plus;

  String get labelKey => 'coach.license.$name';

  static CoachLicenseLevel fromJson(String raw) => CoachLicenseLevel.values
      .firstWhere((c) => c.name == raw, orElse: () => CoachLicenseLevel.dan1);
}

enum CoachSpecialisation {
  kyorugi,
  poomsae,
  technical,
  fitness,
  sparring,
  multi;

  String get labelKey => 'coach.specialisation.$name';

  String get systemIcon => switch (this) {
    CoachSpecialisation.kyorugi => 'figure.martial.arts',
    CoachSpecialisation.poomsae => 'figure.mind.and.body',
    CoachSpecialisation.technical => 'figure.taichi',
    CoachSpecialisation.fitness => 'figure.run',
    CoachSpecialisation.sparring => 'figure.boxing',
    CoachSpecialisation.multi => 'circle.grid.3x3.fill',
  };

  static CoachSpecialisation fromJson(String raw) =>
      CoachSpecialisation.values.firstWhere(
        (c) => c.name == raw,
        orElse: () => CoachSpecialisation.multi,
      );
}

enum CoachEmploymentStatus {
  active,
  leave,
  transferred,
  retired,
  suspended;

  String get labelKey => 'coach.employment.$name';

  static CoachEmploymentStatus fromJson(String raw) =>
      CoachEmploymentStatus.values.firstWhere(
        (c) => c.name == raw,
        orElse: () => CoachEmploymentStatus.active,
      );
}

enum CoachProgramStatus {
  none,
  candidate,
  supportStaff,
  member,
  leadStaff;

  String get labelKey => 'coach.program.$name';

  static CoachProgramStatus fromJson(String raw) => CoachProgramStatus.values
      .firstWhere((c) => c.name == raw, orElse: () => CoachProgramStatus.none);
}

/// Cross-federation ranking snapshot for a coach. Null fields render as "—".
class CoachRanking {
  /// Internal rank inside SSDC (across all coaches).
  final int? club;

  /// UAE Taekwondo Federation coach standing.
  final int? uae;

  /// World Taekwondo recognition tier label (e.g. "Class A", "Class B").
  /// String — WT publishes tiers, not numeric ranks.
  final String? worldTier;

  final DateTime asOf;

  const CoachRanking({this.club, this.uae, this.worldTier, required this.asOf});

  factory CoachRanking.create({
    int? club,
    int? uae,
    String? worldTier,
    DateTime? asOf,
  }) => CoachRanking(
    club: club,
    uae: uae,
    worldTier: worldTier,
    asOf: asOf ?? DateTime.now(),
  );

  factory CoachRanking.fromJson(Map<String, dynamic> json) => CoachRanking(
    club: json['club'] as int?,
    uae: json['uae'] as int?,
    worldTier: json['worldTier'] as String?,
    asOf: DateTime.parse(json['asOf'] as String),
  );

  Map<String, dynamic> toJson() => {
    'club': club,
    'uae': uae,
    'worldTier': worldTier,
    'asOf': asOf.toIso8601String(),
  };
}
